forte/mrdsrg-spin-adapted/sadsrg_comm.cc

```diff
@@ -757,21 +758,192 @@ void SADSRG::V_T2_C2_DF_PH_X(BlockedTensor& B, BlockedTensor& T2, const double&
+     */
 
-    temp = ambit::BlockedTensor::build(tensor_type_, "DFtemp222PHX", jqsb_small);
-    temp["jqsb"] -= alpha * B["gas"] * B["gqm"] * T2["mjba"];
-    temp["jqsb"] -= 0.5 * alpha * L1_["xy"] * T2["yjba"] * B["gas"] * B["gqx"];
-    temp["jqsb"] += 0.5 * alpha * L1_["xy"] * T2["ijbx"] * B["gys"] * B["gqi"];
+    // batch over index f
+    bool batching = false;
+    auto C2blocks = C2.block_labels();
+    for (const std::string& block :
+         {"vcva", "vava", "cvva", "avva", "cvav", "avav", "vcav", "vaav"}) {
+        if (std::find(C2blocks.begin(), C2blocks.end(), block) != C2blocks.end()) {
+            batching = true;
+            break;
+        }
+    }
+    if (batching) {
+        auto nQ = aux_mos_.size();
+        auto nv = virt_mos_.size();
+        auto nc = core_mos_.size();
+        auto na = actv_mos_.size();
+        auto nh = nc + na;
+        auto np = na + nv;
+        auto nvv = nv * nv;
+        auto nav = na * nv;
+        auto ncav = nc * nav;
+        auto naav = na * nav;
+
+        size_t memory_avai = dsrg_mem_.available();
+        memory_avai -= sizeof(double) * (nv * nh * nQ + na * nh * np * np + nh * nh * na * np +
+                                         nv * nh * nv * na * 2);
+        size_t memory_min = sizeof(double) * (nh * nv * nv + nh * nv * na);
+        if (memory_min > memory_avai) {
+            outfile->Printf("\n  Error: Not enough memory for DF-DSRG V_T2_C2_DF_PH_X.");
+            outfile->Printf(" Need at least %zu Bytes more!", memory_min - memory_avai);
+            throw std::runtime_error("Not enough memory to run DF-DSRG V_T2_C2_DF_PH_X!");
+        }
 
-    C2["jqsb"] += temp["jqsb"];
-    C2["qjbs"] += temp["jqsb"];
+        int nthreads = std::min(n_threads_, int(nv));
+        int nthreads_max = static_cast<int>(memory_avai * 0.9 / memory_min);
+        if (nthreads > nthreads_max) {
+            outfile->Printf("\n -> DF-DSRG V_T2_C2_DF_PH_X to be run in %zu threads", nthreads_max);
+            nthreads = nthreads_max;
+        }
+
+        auto Bh = ambit::BlockedTensor::build(tensor_type_, "Bh", {"vhL"});
+        Bh["eig"] = B["gei"];
+        auto Bt = B.block("Lvv");
+
+        auto& Bc_data = Bh.block("vcL").data();
+        auto& Ba_data = Bh.block("vaL").data();
+        auto& Bt_data = Bt.data();
+
+        auto& vcva_data = temp.block("vcva").data();
+        auto& vava_data = temp.block("vava").data();
+        auto& cvva_data = temp.block("cvva").data();
+        auto& avva_data = temp.block("avva").data();
+
+        // temp tensors for each thread
+        std::vector<ambit::BlockedTensor> V_vec(nthreads), Rjew_vec(nthreads);
+        for (int i = 0; i < nthreads; ++i) {
+            std::string t = std::to_string(i);
+            V_vec[i] = ambit::BlockedTensor::build(CoreTensor, "V_thread" + t, {"hvv"});
+            Rjew_vec[i] = ambit::BlockedTensor::build(CoreTensor, "Rjew_thread" + t, {"hva"});
+        }
+
+#pragma omp parallel for num_threads(nthreads)
```

Comment:
Should we leave the scheduling to compiler default (which might be fine if it is `#iterations / #threads` given that this loop looks fairly balanced)?

Reply:
I think the loop is balanced and the default static scheduling should be fine.